modules/rgbd/src/warpfield.cpp

```diff
@@ -0,0 +1,335 @@
+#include "precomp.hpp"
+#include "warpfield.hpp"
+
+namespace cv {
+namespace dynafu {
+
+WarpField::WarpField(int _maxNeighbours, int K, int levels, float baseResolution, float resolutionGrowth):
+k(K), nodes(), maxNeighbours(_maxNeighbours), // good amount for dense kinfu pointclouds
+n_levels(levels), baseRes(baseResolution),
+resGrowthRate(resolutionGrowth),
+regGraphNodes(std::vector<NodeVectorType>(n_levels)),
+nodeIndex(nullptr)
+{
+    CV_Assert(k <= DYNAFU_MAX_NEIGHBOURS);
+}
+
+NodeVectorType WarpField::getNodes() const
+{
+    return nodes;
+}
+
+std::vector<NodeVectorType> WarpField::getGraphNodes() const
+{
+    return regGraphNodes;
+}
+
+bool PtCmp(cv::Point3f a, cv::Point3f b)
+{
+    return (a.x < b.x) ||
+            ((a.x >= b.x) && (a.y < b.y)) ||
+            ((a.x >= b.x) && (a.y >= b.y) && (a.z < b.z));
+}
+
+Ptr<flann::GenericIndex<flann::L2_Simple<float> > > WarpField::getNodeIndex() const
+{
+    return nodeIndex;
+}
+
+Mat getNodesPos(NodeVectorType nv)
+{
+    Mat nodePos((int)nv.size(), 3, CV_32F);
+    for(int i = 0; i < (int)nv.size(); i++)
+    {
+        nodePos.at<float>(i, 0) = nv[i]->pos.x;
+        nodePos.at<float>(i, 1) = nv[i]->pos.y;
+        nodePos.at<float>(i, 2) = nv[i]->pos.z;
+    }
+    return nodePos;
+}
+
+void WarpField::updateNodesFromPoints(InputArray inputPoints)
+{
+    Mat points_matrix(inputPoints.size().height, 3, CV_32F);
+    if(inputPoints.channels() == 1)
+    {
+        points_matrix = inputPoints.getMat().colRange(0, 3);
+    }
+    else
+    {
+        points_matrix = inputPoints.getMat().reshape(1).colRange(0, 3).clone();
+    }
+
+    cvflann::LinearIndexParams params;
+    flann::GenericIndex<flann::L2_Simple <float> > searchIndex(points_matrix, params);
+
+    bool* validIndex;
+    size_t validIndexSize = removeSupported(searchIndex, &validIndex);
+
+    NodeVectorType newNodes;
+    if((int)nodes.size() > k)
+    {
+        newNodes = subsampleIndex(points_matrix, searchIndex, validIndex,
+                                  validIndexSize, baseRes, nodeIndex);
+    }
+    else
+    {
+        newNodes = subsampleIndex(points_matrix, searchIndex, validIndex, validIndexSize, baseRes);
+    }
+
+    initTransforms(newNodes);
+    nodes.insert(nodes.end(), newNodes.begin(), newNodes.end());
+    nodesPos = getNodesPos(nodes);
+    //re-build index
+    nodeIndex = new flann::GenericIndex<flann::L2_Simple<float> >(nodesPos,
+                                                                  cvflann::LinearIndexParams());
+
+    constructRegGraph();
+}
+
+
+size_t WarpField::removeSupported(flann::GenericIndex<flann::L2_Simple<float> >& ind,
+                                  bool** validInd)
+{
+    (*validInd) = new bool[ind.size()];
```

Comment:
1. Where its corresponding `delete`?
2. Why not using `vector`, `AutoBuffer` or any other containter with memory management?

Reply:
1. Looks like I forgot to free the memory.
2. I used a `vector` earlier but each access was taking up too much CPU time on Windows, so I decided to switch. I wasn't aware of `AutoBuffer`, it looks like a better fit in this case.